lib/child_process.js

```diff
@@ -162,24 +170,75 @@ function exec(command, options, callback) {
 }
 
 const customPromiseExecFunction = (orig) => {
-  return (...args) => {
-    let resolve;
-    let reject;
-    const promise = new Promise((res, rej) => {
-      resolve = res;
-      reject = rej;
-    });
+  return (file, options = {}) => {
+    let child;
+
+    if (typeof options !== 'object' || options === null) {
+      return PromiseReject(
+        new ERR_INVALID_ARG_TYPE(
+          'options',
+          'Object',
+          options));
+    }
 
-    promise.child = orig(...args, (err, stdout, stderr) => {
-      if (err !== null) {
-        err.stdout = stdout;
-        err.stderr = stderr;
-        reject(err);
-      } else {
-        resolve({ stdout, stderr });
-      }
+    const { signal } = options;
+    const hasSignal = signal != null;
+
+    // Lets check if an AbortSignal is passed
+    if (hasSignal) {
+      const isNotValidSignal =
+        typeof signal !== 'object' || !('aborted' in signal);
+
+      if (isNotValidSignal) return PromiseReject(
+        new ERR_INVALID_ARG_TYPE(
+          'options.signal',
+          'AbortSignal',
+          signal));
+
+      // Already aborted
+      if (signal.aborted) return PromiseReject(lazyDOMException('AbortError'));
+    }
+
+    const promise = new Promise((resolve, reject) => {
+      if (hasSignal)
+        signal
+          .addEventListener(
+            'abort',
+            rejectOnAbortSignal(reject),
+            { once: true }
+          );
+
+      child = orig(file, options, (err, stdout, stderr) => {
+        if (err !== null) {
```

Comment:
Looking good overall. One additional bit here... once the child process completes, for completeness, there really should be an additional check to see if the `AbortSignal` has been triggered, and if it has, the Promise should be rejected here even tho an error may not have been reported.

Reply:
Got it, in that case, we should do the check even before checking if the `error !== null`, isn't it?